harness/determined/util.py

```diff
@@ -129,48 +129,51 @@ def make_metrics(num_inputs: Optional[int], batch_metrics: List[Dict[str, Any]])
     return metrics
 
 
-def clearinf(obj: Any) -> Any:
-    if isinstance(obj, float) and math.isnan(obj):
-        return "NaN"
-    if obj == math.inf:
-        return "Infinity"
-    if obj == -1 * math.inf:
-        return "-Infinity"
-    if isinstance(obj, List):
-        return list(map(clearinf, obj))
-    if isinstance(obj, dict):
-        return {k: clearinf(v) for k, v in obj.items()}
-    return obj
-
-
 def json_encode(obj: Any, indent: Optional[str] = None, sort_keys: bool = False) -> str:
+    """
+    Encode things as json, with an extra preprocessing step to handle some non-standard types.
+
+    Note: json has a "default" argument that accepts something like our preprocessing step,
+    except it is only invoked for non-native types (i.e. no catching nan or inf floats).
+    """
     import numpy as np
 
-    def json_serializer(obj: Any) -> Any:
+    def jsonable(obj: Any) -> Any:
+        if isinstance(obj, numbers.Integral):
+            # int, np.int64, etc.
+            return int(obj)
+        if isinstance(obj, numbers.Number):
+            obj = cast(SupportsFloat, obj)
+            # float, np.float64, etc.  Serialize nan/±infinity as strings.
+            if math.isfinite(obj):
+                return float(obj)
+            if math.isnan(obj):
+                return "NaN"
+            if obj > 0:
+                return "Infinity"
+            else:
+                return "-Infinity"
```

Comment:
nitpicking: I'd have "infinity" section to look like it's literally checking that the code is dealing with infinity. it's a little weird to have infinity as a last possible option.

```suggestion
            if math.isnan(obj):
                return "NaN"
            if not math.isfinite(obj):
                if obj > 0:
                    return "Infinity"
                else:
                    return "-Infinity"
            return float(obj)
```

Reply:
Stylistically I agree with you.  I was thinking that if this ever had to deal with large arrays, I wanted to pass through the least possible volume of python code.  But probably the difference in performance is negligible.